constraints/src/main/java/gov/nasa/jpl/aerie/constraints/time/IntervalMap.java

```diff
@@ -0,0 +1,437 @@
+        endTime = leftInterval.end;
+        if (leftInterval.includesEnd() && rightInterval.includesEnd()) {
+          endInclusivity = Inclusive;
+          leftIndex = nextLeftIndex;
+          rightIndex = nextRightIndex;
+        } else if (leftInterval.includesEnd()) {
+          endInclusivity = Exclusive;
+          rightIndex = nextRightIndex;
+        } else if (rightInterval.includesEnd()) {
+          endInclusivity = Exclusive;
+          leftIndex = nextLeftIndex;
+        } else {
+          endInclusivity = Exclusive;
+          rightIndex = nextRightIndex;
+        }
+      } else if (leftInterval.end.shorterThan(rightInterval.end)) {
+        endTime = leftInterval.end;
+        endInclusivity = leftInterval.endInclusivity;
+        leftIndex = nextLeftIndex;
+      } else {
+        endTime = rightInterval.end;
+        endInclusivity = rightInterval.endInclusivity;
+        rightIndex = nextRightIndex;
+      }
+      var finalInterval = Interval.between(startTime, startInclusivity, endTime, endInclusivity);
+      if (finalInterval.isEmpty()) continue;
+
+      var newValue = transform.apply(leftValue, rightValue);
+      if (newValue.isPresent()) {
+        if (!newValue.equals(previousValue)) {
+          result.add(Segment.of(finalInterval, newValue.get()));
+        } else {
+          var previousInterval = result.remove(result.size() - 1).interval();
+          result.add(
+              Segment.of(
+                  Interval.unify(previousInterval, finalInterval),
+                  newValue.get()
+              )
+          );
+        }
+      }
+      previousValue = newValue;
+      startTime = endTime;
+      startInclusivity = endInclusivity.opposite();
+    }
+
+    return resultMap;
+  }
+
+  /**
+   * The left bound of the leftmost defined interval.
+   * @return the leftmost bound and its inclusivity
+   */
+  public Optional<Pair<Duration, Interval.Inclusivity>> minValidTimePoint() {
+    if(!this.isEmpty()) {
+      final var window = this.segments.get(0).interval();
+      return Optional.of(Pair.of(window.start, window.startInclusivity));
+    } else{
+      return Optional.empty();
+    }
+  }
+
+  /**
+   * The right bound of the rightmost defined interval.
+   * @return the rightmost bound and its inclusivity
+   */
+  public Optional<Pair<Duration, Interval.Inclusivity>> maxValidTimePoint() {
+    if(!isEmpty()) {
+      final var window = this.segments.get(this.segments.size() - 1).interval();
+      return Optional.of(Pair.of(window.end, window.endInclusivity));
+    } else{
+      return Optional.empty();
+    }
+  }
+
+  private Interval getInterval(final int index) {
+    final var i = (index >= 0) ? index : this.segments.size() - index;
+    return this.segments.get(i).interval();
+  }
+
+  private V getValue(final int index) {
+    final var i = (index >= 0) ? index : this.segments.size() - index;
+    return this.segments.get(i).value();
+  }
+
+  public Segment<V> get(final int index) {
```

Comment:
I noticed that this and the other getters don't do any bounds checking before calling `List.get()`. Is it possible for this method to be called with an out-of-bounds index, and, if so, do we want to do anything in that case?

Reply:
It propagates an error from `List.get()` out to the caller of `IntervalMap.get()`, which has been a helpful behavior during testing.